core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -126,21 +139,23 @@ abstract class AbstractFetcherThread(name: String,
         responseData.foreach { case (topicPartition, partitionData) =>
           val topic = topicPartition.topic
           val partitionId = topicPartition.partition
-          partitionMap.get(topicPartition).foreach(currentPartitionFetchState =>
+          Option(partitionMap.stateValue(topicPartition)).foreach(currentPartitionFetchState =>
             // we append to the log if the current offset is defined and it is the same as the offset requested during fetch
             if (fetchRequest.offset(topicPartition) == currentPartitionFetchState.offset) {
               Errors.forCode(partitionData.errorCode) match {
                 case Errors.NONE =>
                   try {
                     val messages = partitionData.toByteBufferMessageSet
-                    val validBytes = messages.validBytes
                     val newOffset = messages.shallowIterator.toSeq.lastOption match {
-                      case Some(m: MessageAndOffset) => m.nextOffset
-                      case None => currentPartitionFetchState.offset
+                      case Some(m) =>
+                        partitionMap.updateAndMoveToEnd(topicPartition, new PartitionFetchState(m.nextOffset))
```

Comment:
It's a good question. I intend to remove the list as the LinkedHashMap is enough since we always move to the end (initially, I wasn't sure if we may want to keep partitions and topics together in some cases).